### DISCUSSION TITLE
Deserializing subclasses from parse_ methods

### DISCUSSION THREAD

**Discussion by jheddings** [pydantic/pydantic#3091] [Question] [answered]
I've been using the pattern from [PrettyWood](https://github.com/samuelcolvin/pydantic/issues/2177#issuecomment-739578307) to deserialize subclasses.  It has been working very well for members of a model, but I'm unable to use it for parsing top level data.

Note in the output below, `bob` is deserialized properly, but `fluffy` and `tiger` are the generic `Animal` type.  Is there a way to ensure that they are deserialized to the correct subclass?

Also, using `parse_raw_as(Animal, FLUFFY)` works as expected...  Is there a way to get the same behavior using `Animal.parse_raw(FLUFFY)`?

```python
from pydantic import BaseModel

TIGER = """{ "type": "cat", "name": "Tiger the Cat", "color": "tabby" }"""
FLUFFY = """{ "type": "dog", "name": "Fluffy the Dog", "color": "brown", "breed": "rottweiler" }"""

ALICE = """{ "name": "Alice the Person" }"""
BOB = f"""{{ "name": "Bob the Person", "pet": {FLUFFY} }}"""

class Animal(BaseModel):
    type: str
    name: str
    color: str = None

    _subtypes_ = dict()

    def __init_subclass__(cls, type=None):
        cls._subtypes_[type or cls.__name__.lower()] = cls

    @classmethod
    def __get_validators__(cls):
        yield cls._convert_to_real_type_

    @classmethod
    def _convert_to_real_type_(cls, data):
        data_type = data.get("type")

        if data_type is None:
            raise ValueError("Missing 'type' in Animal")

        sub = cls._subtypes_.get(data_type)

        if sub is None:
            raise TypeError(f"Unsupport sub-type: {data_type}")

        return sub(**data)

class Cat(Animal, type="cat"):
    hairless: bool = False

class Dog(Animal, type="dog"):
    breed: str

class Person(BaseModel):
    name: str
    pet: Animal = None

tiger = Animal.parse_raw(TIGER)
print(f"tiger [{tiger.name}] => {type(tiger)}")

fluffy = Animal.parse_raw(FLUFFY)
print(f"fluffy [{fluffy.name}] => {type(fluffy)}")

bob = Person.parse_raw(BOB)
pet = bob.pet
print(f"bob [{bob.name}] => {type(bob)}")
print(f"pet [{pet.name}] => {type(pet)}")
```

Output:

```shell
tiger [Tiger the Cat] => <class '__main__.Animal'>
fluffy [Fluffy the Dog] => <class '__main__.Animal'>
bob [Bob the Person] => <class '__main__.Person'>
pet [Fluffy the Dog] => <class '__main__.Dog'>
```

**Comment by PrettyWood** [comment, marked as answer]
Hi @jheddings 
You can simply add
```py
class Animal(BaseModel):
    ...
    
    @classmethod
    def parse_obj(cls, obj):
        return cls._convert_to_real_type_(obj)
```
If you need to add support for `__root__` and stuff, just have a look at `BaseModel.parse_obj` code.
Hope it helps!

**Reply by jheddings** [reply]
Well, that was surprisingly simple (and now obvious)...  Thanks for the help!

**Comment by MikesRND** [comment]
I am trying to use this coding pattern for serialization/deserialization.  However, I also make frequent use of private attributes.
So I would like `Animal` to inherit from a a baseclass with `underscore_attrs_are_private=True`.  However, this breaks the `_subtypes_` pattern.  I would still like `_subtypes_` to live outside the object dict.  Suggestions appreciated!

```python
class Base(BaseModel)
    class Config:
        underscore_attrs_are_private = True

class Animal(Base):
 ...
```

```
  File "C:/work/examples/pydantic_abc.py", line 42, in <module>
    class Cat(Animal, type="cat"):
  File "pydantic\main.py", line 367, in pydantic.main.ModelMetaclass.__new__
  File "C:\python\Python38\lib\abc.py", line 85, in __new__
    cls = super().__new__(mcls, name, bases, namespace, **kwargs)
  File "C:/work/examples/pydantic_abc.py", line 21, in __init_subclass__
    cls._subtypes_[type or cls.__name__.lower()] = cls
TypeError: 'member_descriptor' object does not support item assignment
```

**Comment by psalvaggio-dl** [comment]
So this solution works great when I am trying to parse the object directly. Is there a way to make this work when the abstract type is included into another pydantic model and `parse_obj` or `parse_raw` or `__init__` is called on that wrapping model?

**Reply by argibbs** [reply]
You mean like in #5785 ?

**Comment by joschu** [comment]
I found this thread and ended up writing a slightly revised (simplified) version, which might be useful for others who stumble across this. Also, curious if there are any problems.

```python

class PolymorphicBase(pydantic.BaseModel):

    # Below code is based on https://github.com/pydantic/pydantic/discussions/3091
    # Enables us to deserialize into the right class given the "type" field

    @classmethod
    def parse_obj(cls, data: dict):
        # Validator might be handed dict or instance of the class
        assert isinstance(data, dict)
        data_type = data.get("type")
        if data_type is None:
            raise ValueError("Need to specify type")
        for sub in cls.__subclasses__():
            if data_type == sub.__fields__["type"].default:
                return sub(**data)
        raise TypeError(f"Unsupport sub-type: {data_type}")

    @classmethod
    def __get_validators__(cls):
        # Otherwise pydantic will try to validate by checking PolymorphicBase fields
        def validate(dict_or_obj):
            if isinstance(dict_or_obj, dict):
                return cls.parse_obj(dict_or_obj)
            elif isinstance(dict_or_obj, cls):
                return dict_or_obj
            else:
                raise TypeError(f"Unsupported type: {type(dict_or_obj)}")

        yield validate
```

**Comment by shakfu** [comment]
Is there a pydantic2 way of doing this? The above methods such as `parse_obj` are all deprecated in pydantic2..

**Reply by samuelcolvin** [reply]
What's "this"? Please give more details about your question.

The short answer is that, yes, it's almost certainly possible with v2, and likely explained in the documentation.

**Reply by shakfu** [reply]
Thanks @samuelcolvin I've given an example on this in this [discussion](https://github.com/pydantic/pydantic/discussions/4456#discussioncomment-7092760) .

Can you please point me to where it is explained in the docs? I haven't been able to find it yet.

**Reply by peter-goldstein** [reply]
Slightly different pattern, but here's what I did:

```python
class Superclass(BaseModel):
  # Already had this function written, might or might not refactor it
  @staticmethod
  def from_dict(obj: Dict[str, Any]) -> "Superclass":
    if "foo" in obj:
      return FooSubclass.model_validate(obj)
    if "bar" in obj:
      return BarSubclass.model_validate(obj)
    raise ValueError()

  @classmethod
  def model_validate(cls, obj, *args, **kwargs):
    # Seems a bit hacky, but this means we only override the model_validate behavior in the superclass
    if cls.__name__ == "Superclass":
      return Superclass.from_dict(obj, *args, **kwargs)
    return super().model_validate(obj, *args, **kwargs)
```

**Reply by florianabel** [reply]
Thanks to everyone providing their solutions so far, very nice approaches.

I am having one doubt though and am wondering if I am thinking it the wrong way.
By implementing it in these ways, won't we require the ParentClass to know about it's ChildClasses in order to perform de/serialization?
Same as using Unions, imho it'd prevent specifically the advantages gained and prevent a Polymorphic approach and the application of dependency inversion.

Still quite new to the party and trying to wrap my head around all the concepts with Pydantic and just wondering how to achieve de/serialization in a way to allow for it. Not sure for the moment if it's possible.

Would be quite happy to receive further hints if I am missing something.

**Reply by tkellogg** [reply]
For pydantic2, [discriminated unions](https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions) may work well if you have a base class with a fixed set of derived classes defined in the same file. That might not work if your inheritance structure is deeper.

Of note, the link above instructs you to list base classes literally in code, if that's not DRY enough for you, you can define a class method like:

```python
class Parent(BaseModel):
    @classmethod
    def get_subclasses(cls):
        return tuple(cls.__subclasses__())
```

And then where you use it, you can pass this result to `Union`:

```python
class User(BaseModel):
    item: Annotated[
        Union[Parent.get_subclasses()],
        Field(discriminator="type"),
    ]
```

Note: `type` is a `Literal` field on every subclass

That code doesn't seem to typecheck on python 3.10, but if you upgrade to python 3.11 you can use a `*` in the union:

```python
        Union[*Parent.get_subclasses()],
```
